src/chain.cpp

```diff
@@ -106,9 +107,9 @@ CBlockIndex* CBlockIndex::GetAncestor(int height)
     return pindexWalk;
 }
 
-const CBlockIndex* CBlockIndex::GetAncestor(int height) const
+CBlockIndex* CBlockIndex::GetAncestor(int height)
 {
-    return const_cast<CBlockIndex*>(this)->GetAncestor(height);
+    return const_cast<CBlockIndex*>(static_cast<const CBlockIndex*>(this)->GetAncestor(height));
```

Comment:
Shouldnt the new cast also be a const_cast?

Reply:
Both `static_cast` and `const_cast` work for casting an object of a type to an object of the same type with added `const`. Personally, I prefer `static_cast` for that case, since `const_cast`, when encountered in code, always has to be examined for its more dangerous use case - removing `const`ness.

Reply:
Interesting, I wasn't aware, but makes perfect sense.

Reply:
Gah, but static_cast could do something crazy like call a constructor or create a temporary. In this case its not like a reader is confused if the resulting type will be const or not.

Reply:
Indeed. So both types of casts here have some disadvantage.
I went to havce look what the C++ Core Guidelines have to say on the matter, and found the exact same usage (with `static_cast`) in their example here: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#example-219

However they list an even better way to do it, which I am going to amend this pull request with :)

Reply:
Gah holy shit lets not use their insane unreadable suggestion there, please. I'd prefer static_cast or const_cast over that.

Reply:
Oh dear, that's the second 'Gah' I've elicited in a single thread :)

Reply:
Heh, I mean I'm OK with static_cast, its not a huge deal...I find const_cast more readable, but its somewhat obvious. The CppCore guidelines there are unreadable garbage, IMO, and would be a step back from the original version.

Reply:
The author of the preferred snippet there is Herb Sutter (commit 2d8377aab56354f7fc9bf759ab134c983a10293e). So in an appeal to authority, I suggest we do use that proposed solution, rather than the multiple casts solution.

Personally, I also think it's a better solution.

Need more opinions here.

Reply:
Please, lets not add a template, decltype gook and a private function just to avoid an obvious const_cast, its not only harder to read but its another function call away instead of direct. Maybe its "the correct way" in that it avoids some casts and magically deduces what you felt like doing, but magic also makes it unreadable to determine exactly what something *does* instead of what it looks like its author probably intended it to do any maybe does.